### PR TITLE
[concurrency] Implement effectful properties

### PR DESCRIPTION
**Pull request by kavon**
https://forums.swift.org/t/pitch-effectful-read-only-properties/44090

TODOs:
- [x] Hide behind `-Xfrontend -enable-experimental-concurrency` flag
- [x] Parsing / serialization / deserialization
  - [x] Implementation
  - [x] Test parsing
  - [x] Test serialization/deserialization.
- [x] Typechecking
  - [x] async / throws read access sites & their coverage
  - [x] Protocols (conformance, composition, inheritance)
  - [x] Class Inheritance & Overrides
  - [x] Actor Isolation
  - [x] Key Paths
  - [x] Prevent annotation with `@objc` (causes crashes in typechecking, for good reason)
  - [x] Make sure protocol test covers subscripts
  - [x] Raise error about effectful props in property wrappers
- [x] Code generation
  - [x] Test SILGen
  - [x] Executable test

Deferred Tasks:
- Property Wrappers
- A useful error message when an attempt is made to use effects in an implicit getter.
- Overload resolution issue in this code, where the `u.someProp` from `A` should be chosen due to being in an `async` context:
```swift
protocol None {
  associatedtype V
  var someProp : V { get }
}
protocol A {
  associatedtype V
  var someProp : V { get async }
}
func f<U : A & None>(u : U) async {
  _ = u.someProp // no errors raised here!
}
```
Note that the checker disallows conformance to `A & None` with an `async` property witness, so this does not ultimately lead to a mis-compile. It is, however, a bit misleading to allow someone to write `f` as it stands, and doesn't match up with how we handle this for functions.

### CODE REVIEW

**Comment by kavon** [comment]
@swift-ci please smoke test

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -4465,6 +4465,24 @@ class AbstractStorageDecl : public ValueDecl {
...
+  /// This check says that \c this is less effectful than \c other if
+  /// \c this either does not exhibit the effect, or if it does, then \c other
+  /// also exhibits the effect.
+  ///
+  /// \param kind the single effect we are performing a check for.
+  ///
+  /// \returns true iff \c this decl has either does not exhibit the effect,
+  ///          or \c other also exhibits the effect.
+  bool isLessEffectfulThan(AbstractStorageDecl const* other,
+                           EffectKind kind) const;
```
Comment (ktoso) [inline include/swift/AST/Decl.h:4484]: :)

File: lib/Sema/TypeCheckProtocol.cpp [hunk 2]
```diff
@@ -488,6 +488,26 @@ matchWitnessDifferentiableAttr(DeclContext *dc, ValueDecl *req,
...
+  if (!witness->isLessEffectfulThan(req, EffectKind::Async))
+    return RequirementMatch(getStandinForAccessor(witness, AccessorKind::Get),
+                            MatchKind::AsyncConflict);
+
+  if (!witness->isLessEffectfulThan(req, EffectKind::Throws))
+    return RequirementMatch(getStandinForAccessor(witness, AccessorKind::Get),
+                            MatchKind::ThrowsConflict);
+
+  return None; // OK
+}
```
Comment (ktoso) [inline lib/Sema/TypeCheckProtocol.cpp:509]: Ah, nice that we also take it into account here in checking protocol conformances. 👍

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (kavon) [reply include/swift/AST/Decl.h:4550]: I just improved this poorly worded comment too :D

**Comment by kavon** [comment]
@swift-ci please smoke test

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Review by DougGregor — APPROVED** [review]
I read through this and don't have any substantial comments, which means I should read through it again... but that shouldn't block merging.

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Comment by kavon** [comment]
This is the PR that I learn about `// REQUIRES: objc_interop` :)

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
Just the driver failure on Linux... will try again because that should be fixed by now.

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by kavon** [comment]
OK I rebased and fixed the change to the asyncHandler note. Rerunning test locally before triggering another run

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Pull request merged**